File: Data Structures in Python/Linked List/SinglyLinkedList.py
class Node:

    def __init__(self):
        self.data = input(print("Enter data: "))
        self.next = None

    def display(self):
        print("Student id = ", self.data)


class linkedList:

    def __init__(self):
        self.start = None

    def getLength(self):
        if self.start is not None:
            count = 1
            ptr = self.start
            while ptr.next is not None:
                count = count + 1
                ptr = ptr.next
            return count
        else:
            print("Linked List is Empty")
            return 0

    def traversing(self):
        ptr = self.start
        if ptr is None:
            print("Empty List ")
        else:
            while ptr is not None:
                print(f"{ptr}\t{ptr.data}\t{ptr.next}")
                # print(ptr.data)
                ptr = ptr.next

    def searching(self, value):
        ptr = self.start
        count = 1
        while ptr is not None:
            if ptr.data == value:
                return count
            else:
                ptr = ptr.next
                count = count + 1
        return -1

    def refresh(self):
        self.start = None
        print("The Linked List has been refreshed by user.")
        
    def insertBegin(self):
        newNode = Node()
        # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
        newNode.next = self.start
        self.start = newNode                        
        # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
        # print(f"{newNode}\t{newNode.data}\t{newNode.next}")

    def insertBeforenode(self, targetNode):
        newNode = Node()
        flag = self.searching(targetNode.data)
        if flag != -1:
            if targetNode != self.start:
                ptr = self.start
                prevPtr = None
                while ptr != targetNode:
                    prevPtr = ptr
                    ptr = ptr.next
                # print(f"{prevPtr}\t{prevPtr.data}\t{prevPtr.next}")
                prevPtr.next = newNode
                # print(f"{prevPtr}\t{prevPtr.data}\t{prevPtr.next}")
                newNode.next = ptr
                # print(f"{newNode}\t{newNode.data}\t{newNode.next}")
            else:
                # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
                newNode.next = self.start
                # print(f"{newNode}\t{newNode.data}\t{newNode.next}")
                self.start = newNode
                # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
        else:
            print("Error: Node to be deleted is not present in the list.")

    def insertAfternode(self, targetNode):
        newNode = Node()
        flag = self.searching(targetNode.data)
        if flag != -1:
            ptr = self.start
            while ptr.next is not None:
                ptr = ptr.next
            if targetNode != ptr:
                ptr = self.start
                nextPtr = ptr.next
                while ptr != targetNode:
                    ptr = ptr.next
                    nextPtr = ptr.next
                # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
                ptr.next = newNode
                # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
                newNode.next = nextPtr
                # print(f"{newNode}\t{newNode.data}\t{newNode.next}")
            else:
                # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
                ptr.next = newNode
                # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
                newNode.next = None    
                # print(f"{newNode}\t{newNode.data}\t{newNode.next}")
        else:
            print("Error: Node to be deleted is not present in the list.")

    def insertEnd(self):
        if self.start is not None:
            newNode = Node()                           
            newNode.next = None
            ptr = self.start
            while ptr.next is not None:
                ptr = ptr.next
            # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
            ptr.next = newNode
            # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
            # print(f"{newNode}\t{newNode.data}\t{newNode.next}")
        else:
            newNode = Node()
            # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
            newNode.next = self.start
            self.start = newNode                        
            # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
            # print(f"{newNode}\t{newNode.data}\t{newNode.next}")

    def delete_startNode(self):
        if self.start is not None:
            # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
            ptr = self.start
            self.start = ptr.next
            # print(f"{self.start}\t{self.start.data}\t{self.start.next}")
            # print(f"{newNode}\t{newNode.data}\t{newNode.next}")
        else:
            print("Error: Linked List is empty and hence, no deletion can be performed.")

    def delete_node_before(self, targetNode):
        print(f"{targetNode}\t{targetNode.data}\t{targetNode.next}")
        flag = self.searching(targetNode.data)
        if flag > 2:
            ptr = self.start
            nextPtr = ptr.next
            while nextPtr.next != targetNode:
                ptr = ptr.next
                nextPtr = ptr.next
            print(f"{ptr}\t{ptr.data}\t{ptr.next}")
            print(f"{nextPtr}\t{nextPtr.data}\t{nextPtr.next}")
            ptr.next = nextPtr.next
            print(f"{ptr}\t{ptr.data}\t{ptr.next}")
            print(f"{nextPtr}\t{nextPtr.data}\t{nextPtr.next}")
        elif flag == 1:
            print("Error: There is no node before this node. Hence, deletion can't be performed.")
        elif flag == 2:
            self.delete_startNode()
        elif flag == -1:
            print("Error: Node to be deleted is not present in the list.")

    def delete_node_after(self, targetNode):
        print(f"{targetNode}\t{targetNode.data}\t{targetNode.next}")
        flag = self.searching(targetNode.data)
        length = self.getLength()
        if flag != -1:
            if flag < length - 1:
                ptr = self.start
                prevPtr = None
                while prevPtr != targetNode:
                    prevPtr = ptr
                    ptr = ptr.next
                # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
                # print(f"{prevPtr}\t{prevPtr.data}\t{prevPtr.next}")
                prevPtr.next = ptr.next
                # print(f"{ptr}\t{ptr.data}\t{ptr.next}")
                # print(f"{prevPtr}\t{prevPtr.data}\t{prevPtr.next}")
            elif flag == length - 1:
                self.delete_endNode()
            elif flag == length:
                print("Error: There is no node present after this. Hence, deletion operation can't be performed.")
        else:
            print("Error: Node to be deleted is not present in the list.")

    def delete_endNode(self):
        if self.start is not None:
            ptr = self.start
            prevPtr = self.start
            while ptr.next is not None:
                prevPtr = ptr
                ptr = ptr.next
            # print(f"{prevPtr}\t{prevPtr.data}\t{prevPtr.next}")
            prevPtr.next = None
            # print(f"{prevPtr}\t{prevPtr.data}\t{prevPtr.next}")
        else:
            print("Error: Node to be deleted is not present in the list.")


node_object = linkedList()
while True:
    choice = int(input(print("Enter your choice:-\n\t1.Insert at start\n\t2.Insert at end\n\t3.Insert before node\n\t4.Insert after node\n\t5.Delete at start\n\t6.Delete at end\n\t7.Delete before node\n\t8.Delete after node\n\t9.Length of linked list\n\t10.Search element\n\t11.Display\n\t12.Refresh\n\t13.Quit"))) 
    if choice == 1:
        node_object.insertBegin()
    elif choice == 2:
        node_object.insertEnd()
    elif choice == 3:
        valNode = input(print("Enter the value at node before which element has to be inserted: "))
        sign = node_object.searching(valNode)
        if sign == -1:
            print("Error: No element with the assigned value exists in the list.")
        else:
            pointer = node_object.start
            for iter in range(sign-1):
                pointer = pointer.next
            node_object.insertBeforenode(pointer)
    elif choice == 4:
        valNode = input(print("Enter the value at node after which element has to be inserted: "))
        sign = node_object.searching(valNode)
        if sign == -1:
            print("Error: No element with the assigned value exists in the list.")
        else:
            pointer = node_object.start
            for iter in range(1, sign):
                pointer = pointer.next
            node_object.insertAfternode(pointer)
    elif choice == 5:
        node_object.delete_startNode()
    elif choice == 6:
        node_object.delete_endNode()
    elif choice == 7:
        valNode = input(print("Enter value at node the element before which is to be deleted : "))
        sign = node_object.searching(valNode)
        if sign == -1:
            print("Error: No element with the assigned value exists in the list.")
        else:
            pointer = node_object.start
            for iter in range(sign-1):
                pointer = pointer.next
            node_object.delete_node_before(pointer)
    elif choice == 8:
        valNode = input(print("Enter value at node the element after which is to be deleted : "))
        sign = node_object.searching(valNode)
        if sign == -1:
            print("Error: No element with the assigned value exists in the list.")
        else:
            pointer = node_object.start
            for iter in range(sign-1):
                pointer = pointer.next
            node_object.delete_node_after(pointer)
    elif choice == 9:
        length_LinkedList = node_object.getLength()
        print(f"Length of Linked List = {length_LinkedList}")
    elif choice == 10:
        valNode = input(print("Enter the value at node which is to be searched: "))
        sign = node_object.searching(valNode)
        if sign == -1:
            print("Error: the entered element is not found in the Linked List")
        else: 
            print(f"Entered element found at {sign}")
    elif choice == 11:
        node_object.traversing()
    elif choice == 12:
        node_object.refresh()
    elif choice == 13:
        exit()
    else:
        print("Invalid choice")

File: Data Structures in C/Linked List/singlyLinkedList.c
// Implementation of Singly Linked List in C Programming Language
#include<stdio.h>
#include<stdlib.h>

#define operations 13

struct node{
    int rollNo;
    struct node* next;
};

struct node* start = NULL;

//Functions to implement a linked list
        //Insertion Operations
struct node* insert_at_begin(struct node* );
struct node* insert_at_end(struct node* );
struct node* insert_before_node(struct node* );
struct node* insert_after_node(struct node* );
        //Deletion Operations
struct node* delete_begin(struct node* );
struct node* delete_end(struct node* );
struct node* delete_before_node(struct node* );
struct node* delete_after_node(struct node* );
struct node* destroy_list(struct node*);
        //Traversing List
void display_list(struct node* );
int searching(struct node* , int );
int length_of_list(struct node* );

int main(){
    printf("\t\t\t\tSINGLY LINKED LIST in C IMPLEMENTATION\n");
    int option, index, num, length;
    do{
    printf("\n********** MAIN MENU **********");
    printf("\n\t1. Insert node at begining");
    printf("\n\t2. Insert node at end");
    printf("\n\t3. Insert node before given node");
    printf("\n\t4. Insert node after given node");
    printf("\n\t5. Delete node at Begining");
    printf("\n\t6. Delete node at End");
    printf("\n\t7. Delete node before given Node");
    printf("\n\t8. Delete node after given Node");
    printf("\n\t9. Search Node in the Linked List");
    printf("\n\t10. Display Linked List");
    printf("\n\t11. Destroy Linked List");
    printf("\n\t12. Length of Linked List");
    printf("\n\t13. Exit Program");
    printf("\nEnter your choice:- ");
    scanf("%d", &option);
    if(option < operations){
        switch(option){
            case 1:
                start = insert_at_begin(start);
                break;
            case 2:
                start = insert_at_end(start);
                break;
            case 3:
                start = insert_before_node(start);
                break;
            case 4:
                start = insert_after_node(start);
                break;
            case 5:
                start = delete_begin(start);
                break;
            case 6:
                start = delete_end(start);
                break; 
            case 7:
                start = delete_before_node(start);
                break;
            case 8:
                start = delete_after_node(start);
                break;
            case 9:
                printf("Enter Roll No. to be searched:- ");
                scanf("%d", &num);
                index = searching(start, num);
                if(index == -1)
                    printf("\nSearch Result:- Entered Node is not present in the Linked List\n");
                else
                    printf("\nSearch Result:- Entered Node is present in the Linked List at index %d\n", index);
                break;
            case 10:
                display_list(start);
                break;
            case 11:
                start = destroy_list(start);
                break;
            case 12:
                length = length_of_list(start);
                printf("\nLength of List = %d\n", length);
                break;
        }
    }
    else if(option == operations){
        start = destroy_list(start);
        printf("\n*********** END OF PROGRAM ***********");
    }
    else
        printf("\nError:- Invalid operation!!!");
    }while(option != operations);
    return 0;
}

struct node* insert_at_begin(struct node* head){
    //printf("\nInserting Node at the Start of List...");
    struct node* newNode;
    int data; 
    printf("\nEnter data for node:- ");
    scanf("%d", &data);
    newNode = (struct node*) malloc(sizeof(struct node));
    if (newNode == NULL)
        printf("\nError:- coludn't allocate space for the node!!!");
    else{
        printf("\nSuccess:- Memory Allocated Successfully...");    
        newNode->rollNo = data;
        newNode->next = head;
        head = newNode;
        //printf("\nNewNode:- Roll No. = %d\tNext Node = %x", newNode->rollNo, newNode->next);
    }
    return head;
}

struct node* insert_before_node(struct node* head){
    //printf("\nInserting Node Before given Node...");
    struct node* newNode;
    struct node *ptr, *prevPtr;
    int data, val;
    printf("\nEnter the Roll No. before which node is to be inserted:- ");
    scanf("%d", &val);
    printf("\nEnter data for the node:- ");
    scanf("%d", &data);
    newNode = (struct node*) malloc(sizeof(struct node));
    if(newNode != NULL){
        printf("\nSuccess:- Memory Allocated Successfully...");    
        newNode->rollNo = data;
        if(head->rollNo == val){
            newNode->next = head;
            head = newNode;
        }
        else{
            ptr = start;
            while(ptr->rollNo != val){
                //printf("\nTraversing through the List...");
                prevPtr = ptr;
                ptr = ptr->next;
            }
            prevPtr->next = newNode;
            newNode->next = ptr;
        }
    }
    else
        printf("\nError:- Couldn't allocate space for node...");
    return head;
}

struct node* insert_after_node(struct node* head){
    printf("\nInserting Node After given Node...");
    struct node* newNode;
    struct node *ptr, *nextPtr;
    int data, val;
    printf("\nEnter the Roll No. after which node is to be inserted:- ");
    scanf("%d", &val);
    printf("\nEnter data for the node:- ");
    scanf("%d", &data);
    newNode = (struct node*) malloc(sizeof(struct node));
    if(newNode != NULL){
        printf("\nSuccess:- Memory Allocated Successfully...\n");
        newNode->rollNo = data;
        ptr = start;
        nextPtr = start->next;
        while(ptr->rollNo != val){
            //printf("Traversing through the List...\n");
            ptr = ptr->next;
            nextPtr = ptr->next;
        }
        ptr->next = newNode;
        newNode->next = nextPtr;
    }
    else
        printf("\nError:- Couldn't allocate space for node...");
    return head;
}

struct node* insert_at_end(struct node* head){
    printf("\nInserting Node at the End...");
    struct node *ptr, *newNode;
    int data;
    
    if(head == NULL)
        head = insert_at_begin(head);
    else{
        printf("\nSuccess:- Memory Allocated Successfully...");    
        printf("\nEnter the data : ");
        scanf("%d", &data);
        newNode = (struct node *)malloc(sizeof(struct node));
        if(newNode == NULL)
            printf("\nError:- Couldn't Allocate Memory");
        else{
            newNode->rollNo = data;
            newNode->next = NULL;
            ptr = start;
            while(ptr->next != NULL)
                ptr = ptr -> next;
            ptr->next = newNode;
            printf("\nNewNode:- Roll No. = %d\tNext Node = %x", newNode->rollNo, newNode->next);
        }
    }
    return head;
}


struct node* delete_begin(struct node* head){
    printf("\nDeleting the first Node...\n");
    struct node* temp = head;
    if(head != NULL){
        temp = head;
        head = head->next;
        printf("\nDeleted Node:- Data = %d   Address = %x", temp->rollNo, temp->next);
        free(temp);
    }
    else
        printf("\nError:- Can't delete from an EMPTY Linked List");
    return head;
}

struct node* delete_before_node(struct node* head){
    //printf("\nDeleting Node Before given Node...\n");
    struct node* ptr;
    struct node* nextPtr;
    struct node* prevPtr;
    int data;
    if(start != NULL){
        printf("\nEnter the Roll No. of the node before which deletion is to take place: ");
        scanf("%d", &data);
        if(head->rollNo == data)
            printf("\nError:- No Node before this node. Deletion can't be performed...");
        else{
            ptr = start;
            nextPtr = start->next;
            while(nextPtr->rollNo != data){
                prevPtr = ptr;
                ptr = ptr->next;
                nextPtr = ptr->next;
            }
            //printf("\nPrevPtr-> %d %x\nPtr-> %d %x\nNextPtr-> %d %x", prevPtr->rollNo, prevPtr->next, ptr->rollNo, ptr->next, nextPtr->rollNo, nextPtr->next);
            prevPtr->next = nextPtr;
            free(ptr);
        }
    }
    else
        printf("\nError:- EMPTY Linked List...\n");
    return head;
}

struct node* delete_after_node(struct node* head){
    //printf("\nDeleting Node After given Node...\n");
    struct node* ptr;
    struct node* last;
    struct node* nextPtr;
    int data;
    if(start != NULL){
        printf("\nEnter the Roll No. of the node before which deletion is to take place: ");
        scanf("%d", &data);
        last = start;
        while(last->next != NULL)
            last = last->next;
        if(last->rollNo == data){
            printf("\nError:-  No Node after this node. Deletion can't be performed...");
        }
        else{   
            ptr = start;
            nextPtr = start->next;
            while(ptr->rollNo != data){
                ptr = ptr->next;
                nextPtr = ptr->next;
            }
            //printf("\nPtr-> %d %x\nNextPtr-> %d %x", ptr->rollNo, ptr->next, nextPtr->rollNo, nextPtr->next);
            ptr->next = nextPtr->next;
            free(nextPtr);
        }
    }
    else
        printf("\nError:- EMPTY Linked List...");
    return head;
}

struct node* delete_end(struct node* head){
    printf("\nDeleting the last Node...\n");
    struct node* ptr;
    struct node* prev;
    if(head != NULL){
        ptr = start;
        while(ptr->next != NULL){
            prev = ptr;
            ptr = ptr->next;
        }
        prev->next = NULL;
        printf("Deleted Node:- Data = %d   Address = %x\n", ptr->rollNo, ptr->next);
        free(ptr); 
    }
    else
        printf("Error:- EMPTY Linked List!!!\n");
    return head;
}

struct node* destroy_list(struct node* head){
    printf("Destroying the entire list...\n");
    struct node* current;
    struct node* nextPointer;
    current = head;
    if(head != NULL){
        while(current != NULL){
            //printf("\nDeleting Node with Roll No. %d ...\n", current->rollNo);
            nextPointer = current->next;
            free(current);
            current = nextPointer;
        }
        head = current;
    }
    else
        printf("Error:- the list is already EMPTY...");
    return head;
}

void display_list(struct node* head){
    printf("\nDisplaying the Linked List...\n");
    struct node* ptr = head;
    if(head == NULL)
        printf("Error:- EMPTY LIST...\n");
    else{
        while(ptr != NULL){
            printf("Current Node:- Data = %d & Address = %x\tNext Node = %x\n", ptr->rollNo, ptr, ptr->next);
            ptr = ptr->next;
        }
    } 
}

// Not Working
int searching(struct node* head, int key){
    printf("\nSearching for Node in the Singly Linked List...");
    //printf("\nkey= %d", key);
    if(head != NULL){
        struct node* ptr;
        ptr = head;
        //printf("\nPtr-> %d %x", ptr->rollNo, ptr->next, ptr);
        int i = 0;
        int test;
        while(ptr != NULL){
            test = (ptr->rollNo == key);
            //printf("\ntest = %d", test);
            //printf("\nIteration = %d\tPtr-> %d %x", i, ptr->rollNo, ptr->next, ptr);
            if(ptr->rollNo == key){
                //printf("inside if-> key = %d", i);
                return i+1;
            }
            ptr = ptr->next;
            i++;
        }
        return -1;
    }
    else{
        printf("Error:- EMPTY List...\n");
        return -1;
    }
}

int length_of_list(struct node* head){
    printf("\nCouting nodes of the Linked List...");
    if(head == NULL)
        return 0;
    else{
        struct node* ptr;
        int len;
        ptr = start;
        for(len=0; ptr != NULL; len++){
            ptr = ptr->next;
        }
        return len;
    }
}
